### ISSUE TITLE
Port pretty printer from PyIRC 2

### ISSUE THREAD

**Issue by Elizafox** [Elizafox/PyIRC#8]
This may or may not belong as an extension, but it would be cool! It'd certainly make a bot's output look a hell of a lot better.

**Comment by Elizafox** [comment]
@awilfox you mentioned you wanted to tackle this?

**Comment by awilfox** [comment]
I actually have the old pretty printer up in a tab.  I was looking at it tonight.  You said you didn't want it as an extension, so where would you want it to live?  Should formatting become a package instead of module?

**Comment by Elizafox** [comment]
I think it could be an optional mixin for IRCBase or so, it seems inappropriate as an extension. What do you think?

**Comment by awilfox** [comment]
My thought is basically:

formatting.py -> formatting/{colours,formatters}.py
Pretty printer -> formatting/pprint.py

**Comment by Elizafox** [comment]
Alright, that works. I'll assign it to you. Feel free to ping me about anything you have questions on.

**Comment by awilfox** [comment]
* Should this be using signals?  Since it isn't an extension, it can't use ```@event```.
* What should it do with unknown lines?
* Should messages that most users won't care about (001-004, CAP, etc) be hidden?  Should that be an option (```hide_fluff``` or such)?

I have code that I could put in a branch if you want to review what I have.

**Comment by Elizafox** [comment]
1. Is it a mixin? IRCBase can use `@event`.
2. If it doesn't know, display it like Konversation. `[numeric] params...`
3. If the user won't care, block it out. Why show it?

**Comment by awilfox** [comment]
1. Yes, it is a mixin.  I didn't realise IRCBase could use ```@event```.
2. Well, since your answer to 1... well... how will I know if it wasn't handled or not?  Do I need to do some deep voodoo involving whether any handler exists for the specific command?
3. Debugging purposes (SASL is a CAP, for instance).

**Comment by Elizafox** [comment]
Hmm... although this sucks, I'd just keep a set of values you do handle and check for membership in that set. I could also add something to PyIRC.signal to allow retrieval of all the stuff attached with `bind`.

**Comment by awilfox** [comment]
Yeah, that's what I've done right now:

```py
        self.handled = ['PRIVMSG', 'NOTICE', 'JOIN', 'PART', 'KICK', '375',
                        '376', '372',]
```

It's just going to be a bit unwieldy...

**Comment by Elizafox** [comment]
See b32bfbb36bb94697a1df74e32dba9664f547ec40. "It may help you on your quest."

**Comment by awilfox** [comment]
This is implemented in 940ea4f!
